futures-util/src/stream/next_some.rs

```diff
@@ -0,0 +1,41 @@
+use core::marker::Unpin;
+use core::pin::Pin;
+use futures_core::stream::{Stream, FusedStream};
+use futures_core::future::{Future, FusedFuture};
+use futures_core::task::{LocalWaker, Poll};
+use crate::stream::StreamExt;
+
+/// A future that resolves to the next value yielded from a [`Stream`].
+#[derive(Debug)]
+#[must_use = "futures do nothing unless polled"]
+pub struct NextSome<'a, St> {
+    stream: Option<&'a mut St>,
+}
+
+impl<'a, St> NextSome<'a, St> {
+    pub(super) fn new(stream: &'a mut St) -> Self {
+        NextSome { stream: Some(stream) }
+    }
+}
+
+impl<'a, St: FusedStream> FusedFuture for NextSome<'a, St> {
+    fn is_terminated(&self) -> bool {
+        self.stream.as_ref().map(|s| s.is_terminated()).unwrap_or(true)
+    }
+}
+
+impl<'a, St: Stream + Unpin> Future for NextSome<'a, St> {
+    type Output = St::Item;
+
+    fn poll(mut self: Pin<&mut Self>, lw: &LocalWaker) -> Poll<Self::Output> {
+        let stream = self.stream.as_mut().expect("NextSome polled after terminated");
+
+        if let Some(item) = ready!(stream.poll_next_unpin(lw)) {
+            Poll::Ready(item)
+        } else {
+            self.stream = None;
+            lw.wake();
```

Comment:
We don't want them to poll, but we do want them to check `is_terminated`. This should deadlock without it when it reaches the end of a stream, but I can't seem to make that happen in a test case...

Reply:
Yeah, I responded below.

Reply:
Actually, I did manage to come up with a test case that shows this is a deadlock without the waking:

```rust
llet mut future = async {                                                                                                                                                              let mut tasks = stream::iter(vec![5]).fuse();
    loop {
        select! {
            num = tasks.next_some() => {
            }
            complete => break,
        }
    }
};
assert_eq!(futures::executor::block_on(future), ());
```

My previous attempts were all using `buffer_unordered` and didn't deadlock, it seems like it might trigger some spurious wakeups.

Reply:
Yup, that makes sense.